napari/_vispy/overlays/scale_bar.py

```diff
@@ -65,13 +67,36 @@ def _calculate_best_length(self, desired_length: float):
         # that might have occurred (e.g. um -> cm)
         factor = current_quantity / new_quantity
 
-        # select value closest to one of our preferred values
-        index = bisect.bisect_left(PREFERRED_VALUES, new_quantity.magnitude)
-        if index > 0:
-            # When we get the lowest index of the list, removing -1 will
-            # return the last index.
-            index -= 1
-        new_value = PREFERRED_VALUES[index]
+        # select value closest to one of our preferred values and also
+        # validate if quantity is dimensionless to prevent the scale bar to
+        # extend beyond the canvas when zooming. If that is the case, calculate
+        # scaled magnitude to use one of the prefered values but scaled
+        # See https://github.com/napari/napari/issues/5914
+        if not new_quantity.dimensionless:
```

Comment:
Is the if/else necessary ? I think the else would work in all cases. I don't know which we want to chose between simplicty and performance

Reply:
Actually, I think this validation can go along side the check about the magnitude being < 1 . Will remove this if/else and change the validation below from: 

`if new_quantity.magnitude < 1` 

to be

`if new_quantity.dimensionless and new_quantity.magnitude < 1`

With that, the logic to parse the new value to a float will only be done when needed and we simplify a little bit the code :+1: